src/main/java/com/google/api/gax/retrying/RetryAlgorithm.java

```diff
@@ -0,0 +1,98 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.retrying;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * The retry algorithm, which makes decision based on the thrown exception and execution time
+ * settings of the previous attempt.
+ *
+ * This class is thread-safe.
+ */
+public class RetryAlgorithm {
+  private final TimedRetryAlgorithm timedAlgorithm;
+  private final ExceptionRetryAlgorithm exceptionAlgorithm;
+
+  /**
+   * Creates a new retry algorithm instance, which uses {@code exceptionAlgorithm} and
+   * {@code timedAlgorithm} to make a decision. {@code exceptionAlgorithm} has higher priority than
+   * the {@code timedAlgorithm}.
+   *
+   * @param timedAlgorithm timed algorithm to use
+   * @param exceptionAlgorithm exception algorithm to use
+   */
+  public RetryAlgorithm(
+      ExceptionRetryAlgorithm exceptionAlgorithm, TimedRetryAlgorithm timedAlgorithm) {
+    this.timedAlgorithm = checkNotNull(timedAlgorithm);
+    this.exceptionAlgorithm = checkNotNull(exceptionAlgorithm);
+  }
+
+  /**
+   * Creates a first attempt {@link TimedAttemptSettings}.
+   *
+   * @return first attempt settings
+   */
+  public TimedAttemptSettings createFirstAttempt() {
+    return timedAlgorithm.createFirstAttempt();
+  }
+
+  /**
+   * Creates a next attempt {@link TimedAttemptSettings}. This method will return the
+   * exception-specific next attempt settings, if there are any, otherwise it will default to the
+   * time-specific settings.
+   *
+   * @param prevThrowable exception thrown by the previous attempt
+   * @param prevSettings previous attempt settings
+   * @return next attempt settings or {@code prevSettings}, if the implementing algorithm does not
+   * provide specific settings for the next attempt
+   */
+  public TimedAttemptSettings createNextAttempt(
+      Throwable prevThrowable, TimedAttemptSettings prevSettings) {
+    TimedAttemptSettings newSettings =
+        exceptionAlgorithm.createNextAttempt(prevThrowable, prevSettings);
+    if (prevSettings.equals(newSettings)) {
```

Comment:
I am against it. It essentially moves us back to what it was before this change - time and exception logics coupled together. If exception algorithm needs information from time alrorithm - this means it does care about timing more than it should. In such cases, I would encourage the clients to write a class which implements both TimedRetryAlgorithm and ExceptionRetryAlgorithm (the implementation may want to aggregate ExponentialRetryAlgorithm if it wants) - because what they need is the tightly coupled time-exception algorithm.

I think we reached good balance between flexibility and complexity of this. Doing more feels like overengineering.